RFC: slot children render function support (#27164)
microsoft/fluentui: rfcs/react-components/convergence/slot-children-render-function.md

On the `AccordionHeader` example, the `button` slot external properties may be found as `props.button` inside the component's
implementation.

`props.button` is called a shorthand and can be either:

1. External properties of the slot
2. `string`, `number`, `Array` or `JSX element` value that will be provided to slot's `children`
3. `null` or `undefined`

```ts
export const AccordionHeader: ForwardRefComponent<AccordionHeaderProps> = React.forwardRef((props, ref) => {
  props.button; // this is the button shorthand
});
```

### Internal properties

Internal properties (`defaultProps`) are the properties provided by the `useState` hook, that should be overridden by
External properties, those are the basic properties that ensure proper functioning of a slot but allowing the user
to opt out of them.

They can be declared when resolving the provided shorthand of a slot (`props.button`, in this example):

```ts
// useAccordionHeader
const buttonShorthand = useARIAButtonShorthand(props.button, {
  required: true,
  // internal properties
  defaultProps: {
    disabled,
    disabledFocusable,
    'aria-expanded': open,
    type: 'button',
  },
});
```

### Overrides

> Overrides are the properties provided by the `render` method, that should override all the other properties.

Author: Sure, but at render time, the actual value that **will** override all the others is the last one provided for the jsx declaration: 

```tsx
// this is the last override before `React.createElement` invocation
<Component {...props} prop={value}/>
```

> Overrides are the properties provided by the `render` method, that should override all the other properties.

Reviewer: Indeed, but we don't pass `props` in `render*()` functions, props are passed via `slotProps`, correct? The single exception is `children` that causing the problem.

wrapping other slots inside of them.

## Current issues

[Slot children render function won't override existing children. #27089](https://github.com/microsoft/fluentui/issues/27089)

Repro: https://codesandbox.io/s/elated-babbage-46ez1w?file=/example.tsx

It's not possible to use children render function to override an entire slot, if `children` are already used. The above is a minimal repro of an issue that can be seen in Fluent in the `AccordionHeader` component. The component renders the `button` slot which contains its own children. JSX children will win over props children - which is how children render functions are rendered by `getSlots`

https://github.com/microsoft/fluentui/blob/220b321042b4bcc126c86a91811a34b04934d49b/packages/react-components/react-accordion/src/components/AccordionHeader/renderAccordionHeader.tsx#L13-L25

Requirements for children render function:

- should be able to override the entire slot
- When rendering the original children, should respect nested slots.

## Problem statement

The main problem revolves around coalescing those 3 instances of properties a slot might have! At the moment,

> our current slot declaration mechanism revolves around 2 separate methods used in 2 different moments:

Author: > does not work in a case when a render template has explicit children

It doesn't work on the case for implicit children also. take `expandIcon` slot on `AccordionHeader` as an example, we don't have an explicit children on the render template, but one is provided by `defaultProps` and is lost since is competing for the same place as the render function itself.

https://github.com/microsoft/fluentui/blob/71b00883c2c3080f13232cbcea92105cf4d70d54/packages/react-components/react-accordion/src/components/AccordionHeader/useAccordionHeader.tsx#L67-L73

element to be render and properties to be passed. In this `getSlots` method, we also ensure that if a given resolved shorthand `children`
property is a function, than we'll invoke that function:

```ts
// getSlots

if (typeof children === 'function') {
  const render = children;
  // in this case the pair of element to be render and properties to be passed are converted into
  // a Fragment and {children} where `children` will be the return of invoking the `children` render method
  return [
    React.Fragment,
    {
      children: render(slot, propsWithoutAsAndChildren),
    } as unknown as R[K],
  ];
}
```

### Problem 1

The `slot` argument will represent the merging between `state.component.button` value
(back into the `AccordionHeader` `button` slot example) and `props.button.as` value. `propsWithoutAsAndChildren`
as the name suggests is all the properties provided by the `resolveShorthand` invocation without including `as` and also `children`
(since `children` by itself is the render function).

Since we lost the Internal properties on the merging between it and the External ones

> There's no way to provide such the default declared `children` to the render function!

Author: Yes, in this case we'll lose `defaultProps.children` which could be the argument to be passed down the line to the render function

> Since we lost the Internal properties on the merging between it and the External ones
> There's no way to provide such the default declared `children` to the render function!

Reviewer: ````suggestion
Since we lost the Internal properties on the merging between it and the External ones
There's no way to provide such the default declared `children` to the render function!

```js
function useComponent(props) {
  return {
    slot: resolveShorthand(props.slot, { children: 'foo' })
  }
}

const state = useComponent({ children: () => {} })
typeof state.slot.children === 'function' // true

// 💥 We lost `defaultProps.children` i.e. `children: "foo"`
```
````

Can you please add that code snippet to the RFC?

### Problem 2

At `render` method after `getSlots` invocation we have all `slots` and `slotProps` well defined. For the case of wrappers
we'll also provide an override for `children`, taking precedence over External properties and also Internal properties.

> This will cause us to lose any children provided by a `render` function passed as a children.

Author: Yes. That's exactly the case for wrappers at the moment.

> This will cause us to lose any children provided by a `render` function passed as a children.

Reviewer: ````suggestion
This will cause us to lose any children provided by a `render` function passed as a children.

```js
function renderComponent(state) {
  const { slots, slotProps } = getSlots(state)

  return <slots.root {...slotProps.root}>Foo</slots.root>
}

const template = renderComponent({ root: { children: 'Baz' }})

template.props.children === 'Baz' // false
template.props.children === 'Foo' // true

// 💥 "Foo" wins over input passed by a user
```
````

   ```

Properties will be lost after those steps, mainly `children` will be lost, affecting the render function mechanism.

Merging the properties is a complex scenario that should not be splitted into multiple steps as all properties
must be considered when merging them, since we got peculiar scenarios involving `children` mostly

## Detailed Design or Proposal

### Option A: Custom JSX Pragma

This problem can be solved by not prematurely merging the 3 instances of properties provided to a slot. To do that
we should find a way to delay as long as possible that merging mechanism.

The last provided properties are the Overrides, which are provided on rendering time, when a jsx element is presented:

```tsx
// slotProps  and children are the overrides here
<slots.button {...slotProps.button}>Children Override</slots.button>
```

Since we only have access to the Overrides on the declaration of the jsx element itself, the only way to postpone
merging of the properties after we have the Overrides is by changing the intrinsic mechanism of how a jsx element is consumed.

Instead of having a `resolveShorthand` that will return a merge between External properties and Internal properties,

> we can provide a custom exotic component, similar to what happens when `React.memo` or `React.forwardRef` does.

Author: No. React tree will remain the same, we'll just add a thin layer of verification/modification before adding a node to that tree

#### slot method over resolveShorthand

The `slot` method will return a slot component, similar to a component declaration when using `React.memo` or `React.forwardRef`,
and in that slot component declaration, both External properties and Internal properties will coexist without merging,
together with the `componentType` that will work as `state.components.slot` would.

This will be very similar to the previous usage of `resolveShorthand`:

```tsx
const state = {
  root: slot(getNativeElementProps(as || 'div', { ref, ...props }), {
    required: true,
    componentType: 'div',
  }),

>   icon: slot(icon, { componentType: 'div' }),

Author: The `slot` method is not equivalent to `React.createElement` the same way `Rect.memo` or `React.forwardRef` aren't. They create an object to be consumed by `React.createElement` later on.

# RFC: Slot children render function

---

_@bsunderhus @ling1726 @layershifter_

## TL;DR

Slot children render function is a complex API that is not properly supported. To support this the best alternative so far,
is to create a custom JSX pragma to ensure slot declaration will not lose any property
and will be capable of properly rendering with children render function.

## Summary

By our [documentation](https://react.fluentui.dev/?path=/docs/concepts-developer-customizing-components-with-slots--page#replacing-the-entire-slot):

When you pass content or props to a slot,
the component renders the content within a component or element based on the slot type.
If you need to replace the slot's entire content, including the containing element,
pass a render function as the children.

This is an escape hatch in the slots API, so prefer the other techniques whenever possible.

> If you replace the entire slot, very accessibility, layout, and styling still work properly.

Author: Yeah, and it seems to be a typo on the documentation, since I copy pasted from it 😂

There's no way to provide such the default declared `children` to the render function!

### Problem 2

At `render` method after `getSlots` invocation we have all `slots` and `slotProps` well defined. For the case of wrappers
we'll also provide an override for `children`, taking precedence over External properties and also Internal properties.
This will cause us to lose any children provided by a `render` function passed as a children.

### Sum up

At the moment the merging of the 3 instances of properties is done in 3 different steps:

1. `resolveShorthand` will merge External properties with Internal properties (external takes precedence)
2. `getSlots` will **wrongly** filter properties on case of render function as `children`
3. on slot rendering of the `render` method the merged properties provided by
   `getSlots` will be merged with the Overrides (overrides takes precedence),
   once again overriding `children` on the case of `wrappers`

   ```tsx
      // the third step is the render of the slot itself
      <slots.button {...slotProps.button}>
   ```

Properties will be lost after those steps, mainly `children` will be lost, affecting the render function mechanism.

> Merging the properties is a complex scenario that should not be splitted into multiple steps as all properties

Reviewer 2: 👍 for this approach, will also simplify the whole underlying architecture

must be considered when merging them, since we got peculiar scenarios involving `children` mostly

## Detailed Design or Proposal

> ### Option A: Custom JSX Pragma

Reviewer 2: Should  this RFC contain more options ? I dont see any "Option B" etc.  if it's only about Custom pragma can we remove that "option" to not confuse reader?

Author: There are other options, this is not the only possible solution (but it's definitely the most elegant one!)

I'll see to add an alternative option just for the sake of better reasoning

    Exclude<Slots[Key], SlotShorthandValue | (Key extends 'root' ? null : never)>
  >;
};
```

After:

```ts
type ComponentState<Slots extends SlotPropsRecord> = {
  [Key in keyof Slots]: SlotComponent<ExtractSlotProps<Slots[Key]>>;
};
```

#### Styles hooks

A minor modification in style hooks are required. at the moment we mutate the merged properties provided by `resolveShorthand`
to include our custom `className` property.

In this case we have some alternatives:

##### Option 1 mutate external properties

Instead of mutating the merged properties, simply mutate the External properties from the slot component:

Before:

```ts
// useAccordionHeader
export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
  const styles = useStyles();
  state.root.className = mergeClasses(
    accordionHeaderClassNames.root,
    styles.root,
    state.inline && styles.rootInline,
    state.disabled && styles.rootDisabled,
    state.root.className,
  );

  return state;
};
```

After:

```ts
// useAccordionHeader
export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
  const styles = useStyles();

>   state.root.props.className = mergeClasses(

Reviewer 2: maybe using a diff would be more appropriate here as there is too much code with subtle differences that can be missed by reader ( I got lost myself :) )

    accordionHeaderClassNames.root,
    styles.root,
    state.inline && styles.rootInline,
    state.disabled && styles.rootDisabled,
    state.root.props.className,
  );

  return state;
};
```

##### Option 2 mutate overrides

> By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can me mutate

Reviewer 2: ```suggestion
By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can be mutated
```

> After:

Reviewer 2: where did "Before" go ?

```tsx
export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
  const styles = useStyles();
  state.overrides.root.className = mergeClasses(
    accordionHeaderClassNames.root,
    styles.root,
    state.inline && styles.rootInline,
    state.disabled && styles.rootDisabled,
    state.root.props.className,
  );

  return state;
};

export const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
  <state.root {...state.overrides.root}>
    <state.button>
      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
      {state.icon && <state.icon />}
      {state.root.props.children}
      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
    </state.button>
  </state.root>
);
```

> ##### Option 3 stop mutating

Reviewer 2: I like this approach most. 

Pros: better DX and in general more uniform react "way" how to approach things
Cons: probably slower  re-render time as creation of immutable non complex structures consume more memory

Author: I'd love to see this approach approved ❤️

For example, on the case `AccordionHeader` the `button` slot would override the `onClick` method. For that a property called `overrides`
can be explicitly added to `AccordionHeaderState` stating such override:

```tsx
// AccordionHeader.types.ts
type AccordionHeaderState = NextComponentState<AccordionHeaderSlots> & {
  overrides: {
    button: {
      onClick(ev: React.MouseEvent<HTMLButtonElement>): void;
    };
  };
};

// useAccordionHeader.ts
const state = {
  // ...
  overrides: {
    button: {
      onClick: useEventCallback(ev => {
        buttonSlot.props.onClick?.(ev);
        if (!ev.defaultPrevented) {
          onAccordionHeaderClick(ev);
        }
      }),
    },
  },
};

// renderAccordionHeader.tsx
const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
  <state.root>
    <state.button onClick={state.overrides.button.onClick}>
      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
      {state.icon && <state.icon />}
      {state.root.props.children}
      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
    </state.button>
  </state.root>
);
```

> #### Custom Pragma implementation

Reviewer 2: how will this affect consumers ? 
- Will they have to switch to use our custom pragma in their codebase ( IMO that's a no go ).

For better context to reader can you please provide more info on how the outputs will look like and what's the consumers story gonna be with this custom pragma ?

Author: > how will this affect consumers ?

It won't (well, it'll allow the children render function support, that's the only difference for the consumers)

> Will they have to switch to use our custom pragma in their codebase ( IMO that's a no go ).

No. JSX pragma  is converted at build time and the consumers will consume the build output with the custom pragma already properly replaced.

> what's the consumers story gonna be with this custom pragma ?

There'll be no difference for consumers story!

> For better context to reader can you please provide more info on how the outputs will look like

👍🏻

    state.inline && styles.rootInline,
    state.disabled && styles.rootDisabled,
-   state.root.className,
+   state.root.props.className,
  );

  return state;
};
```

On the render function side, we just need to spread the overrides:

```tsx
export const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
  <state.root {...state.overrides.root}>
    <state.button>
      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
      {state.icon && <state.icon />}
      {state.root.props.children}
      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
    </state.button>
  </state.root>
);
```

##### Option 3 stop mutating

Let's just stop mutating, and since we require a layer of custom styling through classNames
let's just properly use an argument on the render method exclusively for this:

Before:

```diff
// useAccordionHeader
-export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
+export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState): AccordionHeaderStyles => {
  const styles = useStyles();
+ return {
- state.root.className = mergeClasses(

> +   root: mergeClasses(

Reviewer 3: You only use `root` in the example, but you would need to do the same for every slot right?

Reviewer 3: Could you please update examples to reflect that?

      accordionHeaderClassNames.root,
      styles.root,
      state.inline && styles.rootInline,
      state.disabled && styles.rootDisabled,
-     state.root.className,

> +     state.root.props.className,

Reviewer 3: How would this look in the cases where the style hook needs to set inline styles?

Author: Do we have such case?

Author: Even if we did, then just include `style` and `className` as separate attributes on an object and pass that object as an override to the render method?

Author: I don't mind, I'd just consider an immutable signature

-  <state.root>
+  <state.root className={styles.root}>
    <state.button>
      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
      {state.icon && <state.icon />}
      {state.root.props.children}
      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
    </state.button>
  </state.root>
);
```

#### Passing overrides from state to render

Since our logic lies on the state hooks, the override methods will be implemented on those state hooks, although it should be
declared on the render method explicitly to properly work as an override.

For example, on the case `AccordionHeader` the `button` slot would override the `onClick` method. For that a property called `overrides`
can be explicitly added to `AccordionHeaderState` stating such override:

```tsx
// AccordionHeader.types.ts
type AccordionHeaderState = NextComponentState<AccordionHeaderSlots> & {
  overrides: {
    button: {
      onClick(ev: React.MouseEvent<HTMLButtonElement>): void;
    };
  };
};

// useAccordionHeader.ts
const state = {
  // ...

>   overrides: {

Reviewer 3: why do overrides need to be defined for all slots? it might make sense to define overrides per slot

Author: you don't need to define overrides for every slot. Just for the ones that need overrides.

providing overrides on `slot` is a possible option also!

    button: {
      onClick: useEventCallback(ev => {
        buttonSlot.props.onClick?.(ev);
        if (!ev.defaultPrevented) {
          onAccordionHeaderClick(ev);
        }
      }),
    },
  },
};

// renderAccordionHeader.tsx
const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
  <state.root>

>     <state.button onClick={state.overrides.button.onClick}>

Reviewer 3: would it be possible to spread overrides on slots?

Author: of course

```tsx
<state.button {...state.override.button}>
```

This is also possible

# RFC: Slot children render function

---

_@bsunderhus @ling1726 @layershifter_

## TL;DR

Slot children render function is a complex API that is not properly supported. To support this the best alternative so far,
is to create a custom JSX pragma to ensure slot declaration will not lose any property
and will be capable of properly rendering with children render function.

> ## Summary

Reviewer: Looks that the content below belongs to "Background" section.

> By passing `renderBigLetterIcon` as the `children`, the `span` that normally contains the icon is replaced with an `b` (bold).

```tsx
const renderBigLetterIcon (Component, props) => {
  return <b>B</b>;
};

<Button icon={{ children: renderBigLetterIcon }}>Bold</Button>;
```

## Background

There are 3 instances of properties that will be provided to a slot in different phases of the slot creation:

1. External properties (`props`)
2. Internal properties (`defaultProps`)
3. Overrides

### External properties

External properties (`props`) are the properties provided by the client using a component with a specific slot.
On the `AccordionHeader` example, the `button` slot external properties may be found as `props.button` inside the component's
implementation.

`props.button` is called a shorthand and can be either:

1. External properties of the slot (an object)
2. `string`, `number`, `Array` or `JSX element` value that will be provided to slot's `children`
3. `null` or `undefined`

```ts
export const AccordionHeader: ForwardRefComponent<AccordionHeaderProps> = React.forwardRef((props, ref) => {
  props.button; // this is the button shorthand
});
```

### Internal properties

> Internal properties (`defaultProps`) are the properties provided by the `useState` hook, that should be overridden by

Reviewer: ```suggestion
Internal properties (`defaultProps`) are the properties provided by the `use_COMPONENT_` hook, that should be overridden by
```

?

External properties, those are the basic properties that ensure proper functioning of a slot but allowing the user
to opt out of them.

They can be declared when resolving the provided shorthand of a slot (`props.button`, in this example):

```ts
// useAccordionHeader
const buttonShorthand = useARIAButtonShorthand(props.button, {
  required: true,
  // internal properties
  defaultProps: {

>     disabled,
>     disabledFocusable,

Reviewer: ```suggestion
```

Let's remove them from the example as it's unclear from there `disabled` & `disabledFocusable` are coming

A minimal version of option A, but without introducing internal changes. If we focus on the problem itself,
slot children render function, the main requirement would be to maintain `defaultProps.children` until the last possible
moment so that it can be introduced once again back to a children render function.

This proposal would take advantage of `Symbol` to introduce a "private" method to the result of `resolveShorthand`,
to maintain the value of `defaultProps.children`. That value would then later on be consumed by a minimal custom JSX pragma
that on the specific case of children render function this pragma would consume this `defaultProps.children` to ensure proper functionality.

```diff
const defaultPropsChildrenSymbol = Symbol('fuiSlotDefaultPropsChildren')

export const resolveShorthand: ResolveShorthandFunction = (value, options) => {
  //...

  return defaultProps ? {
    ...defaultProps,
    ...resolvedShorthand,
+   [defaultPropsChildrenSymbol]: defaultProps?.children
  } : resolvedShorthand;
};
```

https://codesandbox.io/s/wispy-leftpad-f8yi37?file=/example.tsx.

#### Pros and Cons

##### Pros

1. No internal changes required
2. No migration is required

##### Cons

> 1. Requires extra build step

Reviewer: I don't think so, we just need to put a magic comment to `render*.tsx` files i.e.

```diff
+/** @jsxRuntime classic */
+/** @jsx fluentCreateElement */

```

Author: yeaaah, I was trying to hide this, as it honestly sounds like a hack to add comments to every single file we want to change pragma. I'd rather spend the time to update our build process to simply include the pragma change, sounds more elegant.

### Option D: Option A + Option C

Taking the strategy provided on Option C to hide required data with `Symbol` inside the result value of `resolveShorthand`
and applying it to not only move forward the `defaultProps.children` (from Option C) but also `componentType` (from Option A)
we can completely get rid of `getSlots` invocation, as all the data required to render a slot will be provided by `resolveShorthand` and only consumed at render by a custom pragma.

```diff
const defaultPropsChildrenSymbol = Symbol('fuiSlotDefaultPropsChildren')
const componentTypeSymbol = Symbol('fuiSlotComponentType')

export const resolveShorthand: ResolveShorthandFunction = (value, options) => {
  //...

  return defaultProps ? {
    ...defaultProps,
    ...resolvedShorthand,
+   [defaultPropsChildrenSymbol]: defaultProps?.children

> +   [componentTypeSymbol]: options.componentType

Reviewer: I would rather make an explicit argument (probably `first`), so the contract will be clearer and we will avoid questions like "why it renders a div?!"

  } : resolvedShorthand;
};
```

##### Pros

1. Simple to use
2. Reduces codebase size (`getSlots` invocations can be stripped)
3. No iteration over `state.components` is required to define slots (the slots used are the slots being processed)
4. Will simplify the whole underlying architecture
5. No Breaking changes, the consumers will not have any impact (besides allowing the children render function to work properly)

> 6. couples together `state.components` with the slot itself

Reviewer: I would say that in this case it's reasonable to get rid from `state.components` at all as I don't see any reasons to keep it.

##### Cons

> 1. Requires extra build steps

Reviewer: The same as for the previous option, I don't think that it requires any build changes.

`getSlots` is invoked at `render` method and will go over every single slot declared on `state.components` and convert every resolved shorthand into a pair of
element to be render and properties to be passed. In this `getSlots` method, we also ensure that if a given resolved shorthand `children`
property is a function, than we'll invoke that function:

```ts
// getSlots

if (typeof children === 'function') {
  const render = children;
  // in this case the pair of element to be render and properties to be passed are converted into
  // a Fragment and {children} where `children` will be the return of invoking the `children` render method
  return [
    React.Fragment,
    {
      children: render(slot, propsWithoutAsAndChildren),
    } as unknown as R[K],
  ];
}
```

### Problem 1

The `slot` argument will represent the merging between `state.component.button` value
(back into the `AccordionHeader` `button` slot example) and `props.button.as` value. `propsWithoutAsAndChildren`
as the name suggests is all the properties provided by the `resolveShorthand` invocation without including `as` and also `children`
(since `children` by itself is the render function).

> Since we lost the Internal properties on the merging between it and the External ones
> There's no way to provide such the default declared `children` to the render function!

Reviewer 4: nit

```suggestion
Since we lost the Internal properties on the merging between it and the external ones,
there's no way to provide such the default declared `children` to the render function!
```

};
```

##### Styles hooks `.props` access

A minor modification in style hooks are required. at the moment we mutate the merged properties provided by `resolveShorthand`
to include our custom `className` property.

In this case we have some alternatives:

###### Option 1 mutate external properties

Instead of mutating the merged properties, simply mutate the External properties from the slot component:

```diff
// useAccordionHeader
export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
  const styles = useStyles();
- state.root.className = mergeClasses(
+ state.root.props.className = mergeClasses(
    accordionHeaderClassNames.root,
    styles.root,
    state.inline && styles.rootInline,
    state.disabled && styles.rootDisabled,
-   state.root.className,
+   state.root.props.className,
  );

  if (state.icon) {
-    state.icon.className = mergeClasses(
+    state.icon.props.className = mergeClasses(
      accordionHeaderClassNames.icon,
      styles.icon,
-      state.icon.className
+      state.icon.props.className
    );
  }

  return state;
};
```

###### Option 2 mutate overrides

> By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can be mutated:

Author: We can always stick with option 1 and just keep mutating `props`. I'm particularly fond of option 3 though....

> By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can be mutated:

Reviewer 5: Yes, I think mutating props is the best option IMO.
* Having separate `overrides` vs. `props` vs. `defaultProps` quickly gets difficult to manage.
* Having immutable `props` and copying the object on every override could have perf issues.

> By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can be mutated:

Author: Technically it would just be 2 objects on overrides, 1 for the styles and 1 for all overrides that come from the state hook, I don't see why more than that would be required, and they don't even need to be copied, we can merge them in the render method.

- state.root.className = mergeClasses(
+ state.overrides.root.className = mergeClasses(
    accordionHeaderClassNames.root,
    styles.root,
    state.inline && styles.rootInline,
    state.disabled && styles.rootDisabled,
-   state.root.className,
+   state.root.props.className,
  );

    if (state.icon) {
-    state.icon.className = mergeClasses(
+    state.overrides.icon.className = mergeClasses(
      accordionHeaderClassNames.icon,
      styles.icon,
-      state.icon.className
+      state.icon.props.className
    );
  }

  return state;
};
```

On the render function side, we just need to spread the overrides:

```tsx
export const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
  <state.root {...state.overrides.root}>
    <state.button>
      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
      {state.icon && <state.icon />}
      {state.root.props.children}
      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
    </state.button>
  </state.root>
);
```

###### Option 3 stop mutating

Let's just stop mutating, and since we require a layer of custom styling through classNames

> let's just properly use an argument on the render method exclusively for this:

Author: But to be clear, the option 3 was focused solely on creating a new signature for the `styles` only, not for all overrides

> let's just properly use an argument on the render method exclusively for this:

Reviewer 5: > The other way to merge the props together would be having a `state.overrides.SLOT` , and then simply adding the overrides on render function.

Like I mentioned in my other comment https://github.com/microsoft/fluentui/pull/27164#discussion_r1149605391, I don't think `state.overrides.SLOT` will end up being a good idea: it gets very tricky to check whether you need to merge with `overrides` vs. `props` vs. `defaultProps`. Instead, I think mutating `state.SLOT.props` is the best option for overrides.

> But to be clear, the option 3 was focused solely on creating a new signature for the `styles` only, not for all overrides

The change to the `useComponentStyles` hooks seems like a separate RFC from this one. It might be better to split that proposal out, in order to keep this one focused on the other changes (JSX pragma, etc.)

For example, on the case `AccordionHeader` the `button` slot would override the `onClick` method. For that a property called `overrides`
can be explicitly added to `AccordionHeaderState` stating such override:

```tsx
// AccordionHeader.types.ts
type AccordionHeaderState = NextComponentState<AccordionHeaderSlots> & {
  overrides: {
    button: {
      onClick(ev: React.MouseEvent<HTMLButtonElement>): void;
    };
  };
};

// useAccordionHeader.ts
const state = {
  // ...
  overrides: {
    button: {
      onClick: useEventCallback(ev => {
        buttonSlot.props.onClick?.(ev);
        if (!ev.defaultPrevented) {
          onAccordionHeaderClick(ev);
        }
      }),
    },
  },
};

// renderAccordionHeader.tsx
const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
  <state.root>
    <state.button onClick={state.overrides.button.onClick}>
      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
      {state.icon && <state.icon />}
      {state.root.props.children}
      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
    </state.button>
  </state.root>
);
```

> #### Custom Pragma implementation

Author: > I think we'd need to have a way to override props for every slot, built-in:
We'll always need to override className.
Someone re-composing our component may want to override props for a slot that we don't override internally.

Totally doable, we could include a `state.overrides.SLOT` as part of the proposal

> #### Custom Pragma implementation

Author: > I think this idea has good merit, and I think we can do without a custom pragma. The JSX pragma is just syntactic sugar, but if we simplify the non-JSX syntax enough, we could skip using JSX in the render functions entirely.
> 
> Here's an idea that might work ok without needing a JSX pragma. A few background assumptions:
> 
> * The only job of the `renderFoo` function is to define the _tree structure_ of the control (i.e. what slots go where in the DOM tree); not to override props or add functionality.
> * We never should have property overrides in `renderFoo`: they should all be handled by the `useFoo` hook(s).
> * All of our render functions adhere to this, and don't provide custom prop overrides in the render function.
> 
> With those assumptions, we could then write a `renderSlot` function that takes the slot object, and children:
> 
> ```ts
> function renderSlot<Props extends UnknownSlotProps>(
>   component: SlotComponent<Props>, 
>   children: React.ReactNode[]
> ): React.ReactElement<Props> | null {
>    // Works just like jsxFromSlotComponent, except without override handling
> }
> ```
> 
> Then a render function might look like:
> 
> ```ts
> const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
>   renderSlot(state.root, [
>     renderSlot(state.button, [
>       state.expandIconPosition === 'start' && state.expandIcon && renderSlot(state.expandIcon),
>       state.icon && renderSlot(state.icon),
>       ...state.root.props.children,
>       state.expandIconPosition === 'end' && state.expandIcon && renderSlot(state.expandIcon),
>     ]),
>   ]);
> );
> ```
> 
> ... a little old school like calling React.createElement manually. The main distinction is that by preventing further overrides of props, the function doesn't become a mess of props mixed with children: it _only_ defines the structure of the slots via the children array.

This is just going back to the way things are done in N*. I believe a custom pragma is a better way forward for readability

> #### Custom Pragma implementation

Author: > We could still wrap all this with a custom JSX pragma if we wanted to. I'm not strongly opposed to it, although it has some downsides: namely developer confusion about why our JSX works differently from others. We could consider whether or not to add a JSX pragma as a separate topic.
> 
> But for reference, with a custom pragma:
> 
> ```ts
> function jsxRenderSlot(
>   component: SlotComponent<Props>,
>   overrides?: never, // prevent any overrides from being added -- not sure if this is the right type
>   ...children: React.ReactNode[]
> ): {
>   return renderSlot(component, children);
> }
> ```
> 
> The result would look more similar to your other examples, except you couldn't specify any extra props on the nodes.
> 
> ```ts
> const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
>   <state.root>
>     <state.button>
>       {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
>       {state.icon && <state.icon />}
>       {state.root.props.children}
>       {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
>     </state.button>
>   </state.root>
> );
> ```

Yeah, that's not the right types but I got what is your idea. So in this ase we would completely avoid overrides from the render methods and we would need to introduce it solely on the state hooks, right?! @behowell, I'm not opposed to this idea, it could blend in just fine with option A, C, D

> #### Custom Pragma implementation

Reviewer: > The custom JSX pragma could be confusing to other developers, since it is a rare and unexpected thing to do.

There is adoption in the community, as it's one of options to use `css` prop with Emotion: https://emotion.sh/docs/css-prop#jsx-pragma

> #### Custom Pragma implementation

Author: > Just wanted to summarize my opinion here:
> 
> * A custom JSX pragma is not _required_ for the changes in this RFC. The changes in this RFC can be implemented with a function like `renderSlot`.

Wasn't this clear already? option B is basically the same idea as this `renderSlot` but focused in the children problem, the main difference is that this `renderSlot` approach just goes way back to the way things were done in N*.

> * IMO the function-only approach is _only slightly_ less readable than the JSX pragma.

I disagree completely, and I believe the whole idea behind the creation of JSX syntax also disagrees with this argument 😅. 

> * We could optionally add the JSX pragma later if we decide to.

Yes, that would be something similar to option B. I guess my point with this RFC was proposing changes **together**
with the solution of the problem. I guess I didn't emphasise that well, I'll try to do it better.
